Minimal interface

kenju opened the pull request:
## About

Publish `ResponsiveImg` as a npm package in a production quality level.

## Not TODO

Maybe in the next PR scpre:

- Setup sample page in github-pages
- Setup example repos
- Setup test coverage

## Links

- http://w3c.github.io/html/semantics-embedded-content.html#the-img-element

kenju committed:
Add package.json

kenju committed:
Add gitignore

kenju committed:
Add packlage.json

kenju committed:
Add .babelrc

kenju committed:
Add yarn.lock

kenju committed:
Copy ResponsiveImg into this repo

kenju committed:
Pass current cli spec

kenju committed:
Remove unnecessary file from git

kenju committed:
Update spec to accept both pixel/width desciptor

kenju committed:
Add sizez test for widthDescriptor

kenju committed:
Add eslint

kenju committed:
Add eslint config file

kenju committed:
Update indent size from 4 to 2

kenju committed:
Fix type

gfx commented:
方針はいいと思います。

ところでcomponentの名前は単に "Image" (`<Image/>`) でいいかなという気がします。

というのも、responsive imageを簡単に作るためのものですが、これ自体なにか特別なことをするわけではなく、HTML5標準の `<img/>` へのインターフェイスにすぎないからです。

パッケージ名も `react-responsive-image` にしましょう。

gfx commented on package.json:
```diff
@@ -0,0 +1,42 @@
+    "power-assert": "^1.4.2",
+    "mocha": "^3.2.0"
+  },
+  "dependencies": {
```
reactへの依存はpeerDependenciesにしましょう。

あとreact-domはコンポーネント内では使ってないですよね。なのでdevDependenciesdえよさそうです。

kenju replied:
peerDependencies だと node_modules/ 配下に react がインストールされないため、testDependenciesにも含めました。

以下のような解決策もあるようです。circle ciのスクリプトにreactをインストールするよう仕込む方法も考えられると思います。
http://developer.hatenastaff.com/entry/2016/04/09/180000#react%E3%81%AFpeerDependencies%E3%81%AB%E5%85%A5%E3%82%8C%E3%82%8B

kenju committed:
Add width/pixel descriptor

kenju committed:
Add width/pixel descriptor

kenju committed:
Fix spec when both descriptor is given

kenju committed:
Run eslint --fix

kenju committed:
Fix eslint warning of arrow function parenthesis

kenju committed:
Fix arrow-spacing

kenju committed:
Fix jsx props indentation

kenju committed:
Reduce redundancy by setting descriptor to state.descriptor

kenju committed:
Fix eslint warning

kenju committed:
s/ResponseImg/Img

kenju committed:
Create sizezShape const

kenju committed:
Move react-dom to devDependencies

kenju committed:
s/ResponsiveImg/Image in spec

kenju committed:
Make react peerDependencies

kenju committed:
s/Img/Image

kenju committed:
circle.yml

kenju committed:
Update circle.yml

kenju committed:
Tweak circle ci config

kenju committed:
Update pretest and posttest spec to install peerDependencies react
@see
http://developer.hatenastaff.com/entry/2016/04/09/180000#reactはpeerDependenciesに入れる

kenju committed:
Fix package.json to include react in devDependencies as well

kenju commented:
> ところでcomponentの名前は単に "Image" (<Image/>) でいいかなという気がします。

たしかにそうですね！シンプルにしました。

@bitjourney/developers 
@gfx 

だいたい実装完了しましたので、インターフェース部分やそのほか気になるところあればフィードバック・コメントいただけますでしょうか？ :pray:

https://github.com/bitjourney/react-responsive-image/pull/1/files#diff-4f4be899712fbaab94366df5ad7259a3

kenju committed:
Add minimal document

kenju committed:
Fix readme codeblock syntax highlight

kenju committed:
Fix package entry point

kenju committed:
Remove redundant circle ci command

kenju committed:
Move proptypes to static field

kenju committed:
Fix eslint warning

kenju committed:
Fix spec eslint

kenju committed:
Add eslint ignore

kenju committed:
Add npmignore

kenju commented:
@gfx 

https://github.com/mikefey/react-responsive-image
react-responsive-image がすでに存在するみたいなので、`react-image-tag` とでもしようと思っているのですが、どう思いますか？

gfx commented on .npmignore:
```diff
@@ -0,0 +1,19 @@
+
+# for editor settings
+.idea/
+.vscode
```
FYI: .vscode/settings.json は入れてしまったほうがいいと思いますね。そのほうがESLintの自動適用を強制できるしコントリビュートする際に気をつけることが減るので。

gfx commented:
FYI: .vscode/settings.json はgit repoに入れてしまったほうがいいと思いますね。そのほうがESLintの自動適用を強制できるしコントリビュートする際に気をつけることが減るので。

gfx commented on README.md:
```diff
@@ -0,0 +1,90 @@
+# react-responsive-image
+
+`react-responsive-image` is a React Components of <img> tag with cleaner srcset/sizes interface.
```
```diff
- is a React Components of <img> tag
+ is a React component of the `<img/>` tag
```

※ markdown内ではHTML tagが有効なことがあるのでbackquoteで囲んだほうがよい

gfx commented:
`react-image-tag` だと何の説明にもなってないのでよくないですね。

`react-simple-image` とかどうでしょう。

gfx commented on README.md:
```diff
@@ -0,0 +1,90 @@
+    srcSet: [
+      {descriptor: '1x', src: 'example.svg'},
+      {descriptor: '2x', src: 'example@2x.svg'},
+    ],
```
なるほどこういう感じにしたんですね。もともとの `"1x" : "example.svg"` でも曖昧ではないからインターフェイスとしては十分であるように思いますが、どういう理由でこういうインターフェイスにしたんですか？

余談ですが、SVGはvector imageなので`@2x`とか用意する必要がないんですよね。なのでサンプルはPNGかJpegにしておきましょう。

kenju replied:
> なるほどこういう感じにしたんですね。もともとの "1x" : "example.svg" でも曖昧ではないからインターフェイスとしては十分であるように思いますが、どういう理由でこういうインターフェイスにしたんですか？

[EXAMPLE11](http://w3c.github.io/html/semantics-embedded-content.html#the-img-element) に書いてあるのですが、1x, 2x以外にも1.5xのような可能性があり得るようなのです。なので、迷った箇所ではあるのですが、`widthDescriptor` と同様のインターフェースに統一した方が全体としてはシンプルかな、と思いました。

```html
<img src="/uploads/100-marie-lloyd.jpg"
      srcset="/uploads/150-marie-lloyd.jpg 1.5x, /uploads/200-marie-lloyd.jpg 2x"
      alt="" width="100" height="150">
```

とはいえこちらのdescriptorの可能性としては、現実問題1x, 1.5x, 2x程度だとは思います。そこをライブラリ側で強制して良いかどうかは迷っています。どう思いますか？

gfx replied:
なるほど、つまり「descriptorは任意の値をとるのでReact shapeで表現できずtype-safeではないから」ということですね。しかしそれにしては `{ descriptor: ..., src: ... }` も別にtype-safeではないんですよね。

というわけで、シンプル形式をとりつつcustom prop callbackでdescriptorの値をvalidateするのはどうでしょうか。
https://facebook.github.io/react/docs/typechecking-with-proptypes.html

何故ここにこだわるかとういと、「srcsetをシンプルかつtype-safeに書けること」がこのreact componentのコアバリューだと思うからです。type-safeにこだわらないなら、はじめから `React.PropTypes.object.isRequired` で十分です。

kenju replied:
> 何故ここにこだわるかとういと、「srcsetをシンプルかつtype-safeに書けること」がこのreact componentのコアバリューだと思うからです。

なるほど。type-safe もコアバリュー。

とすると、 `(\d)+(\.\d+)?(w|x|)` あたりの正規表現で、それぞれのdescriptorが「数値＋`x | w`」であることを保証するカスタムバリデーションを実装してみるのが方針として良さそうです。

kenju committed:
s/svg/png

kenju committed:
s/react-responsive-image/react-simple-image

kenju committed:
Add html markup

kenju committed:
Reduce nested descriptor props

kenju committed:
Add validation logic not to render sizez props not when width descriptor only

kenju committed:
Add should word to spec describe

kenju committed:
Add spec

kenju committed:
Pass ci not to render both pixel and width descriptor

kenju committed:
Fix eslint warning

kenju committed:
Update README

kenju committed:
Add .vscode

kenju commented on src/image.jsx:
```diff
@@ -0,0 +1,72 @@
+      alt: React.PropTypes.string.isRequired,
+      className: React.PropTypes.string,
+      srcSet: React.PropTypes.arrayOf(React.PropTypes.shape({
+        descriptor: (props, propName, componentName) => {
```
https://github.com/bitjourney/react-simple-image/pull/1#discussion_r93838340

@gfx 

↑のコメントでもらったところですが、Custom Validationを用いてこのように実装できました。
`widthDescriptor`, `pixelDescriptor`... というレイヤーはインターフェースから取り払いました。

gfx replied:
いいと思います 👍

gfx replied:
ん？インターフェイスは `{ descriptor: '360w', src: 'example-small.png' },` のままなんですか？ `{ '360w':  'example-small.png' },` ではなく？それはなぜですか？後者のほうがずっとシンプルだと思うのですが。

kenju replied:
最初 key が動的な descriptor の実装のイメージがつかなかったのですね。ですがもう一度考え直したら、keyが動的なObjectは`Object.keys()`で取得できると思い、実装し直しました。

（本来keyが一つのオブジェクトが渡ってくる事前条件をもとにしているので、配列で取得するのは少し違和感を覚えますが、しょうがない）

ちなみに、`sizez`の方は、sizeもmediaConditionもどちらもkeyにふさわしいような値ではないし、今のままで良いかなと思っています。

kenju committed:
Remove comments

kenju committed:
Update README

kenju committed:
Add LICENSE

kenju committed:
Add CHANGES

kenju commented:
@gfx @bitjourney/developers 

インターフェースのリファクタやテストの追加を行いました。よければまた見てアドバイスなどいただけませんでしょうか？ 🙏 

CI設定ファイルなども混じっているので、[README.md](https://github.com/bitjourney/react-simple-image/blob/71b21eb7a8a2eb6114d122772cbb95b16c048eb9/README.md) を最初に見ていただければ、全体感がつかめるかと思います。

LICENSEは、もちろん他のBJレポジトリを見てってのはあるのですが、色々調べてApache 2.0にしました。これについてどう考えたらいいとかアドバイスあればぜひ聞きたいです。

gfx commented on src/image.jsx:
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+
+const REGEXP_DESCRIPTOR_WIDTH = /(\d)+(\.\d+)?w$/;
```
`/^\d+(?:\.\d+)?w$/` にしましょう。なぜこうすべきなのかは『詳説正規表現』を読んでください。

kenju replied:
ここは本読んで出直してきます。一旦TODOです。

gfx replied:
現状だと要件を満たしていないので直すだけなおしておいてください。

たとえば現状は "foo1w" でもマッチします

kenju replied:
もろもろ修正しました！

gfx commented on src/image.jsx:
```diff
@@ -0,0 +1,72 @@
+
+const REGEXP_DESCRIPTOR_WIDTH = /(\d)+(\.\d+)?w$/;
+const REGEXP_DESCRIPTOR_PIXEL = /(\d)+(\.\d+)?x$/;
+const REGEXP_DESCRIPTOR_WIDTH_AND_PIXEL = /(\d)+(\.\d+)?(w|x)$/;
```
`/^\d+(?:\.\d+)?[wx]$/` にしましょう。

kenju replied:
- ^ は凡ミスでしたね。
- ?: はキャプチャなしの括弧で、必要ないグループ化は明示的にキャプチャ不要であることを宣言的に書くべき
- | ではなく [] を用いた選択をするのは、バックトラックの回数が増えることで前者の方がパフォーマンスが高くなってしまうから

理解あってましたでしょうか？

gfx replied:
あってます 👏 

あと `(\d)+` はキャプチャが必要ないならそもそも `\d+` で十分ということですね。

kenju replied:
> あと (\d)+ はキャプチャが必要ないならそもそも \d+ で十分ということですね。

ふむ、なるほど！理解しました 🙋

kenju committed:
Update interface

kenju committed:
Fix eslint

kenju committed:
Add Matcher

kenju committed:
Fix matcher

kenju committed:
Ignore arrow body style

kenju committed:
Add spec for Matcher#matchWidthDescriptor

kenju committed:
Fix width descriptor matcher

kenju committed:
Add spec for Matcher#matchPixelDescriptor

kenju committed:
Add spec for Match#matchDescriptor

kenju committed:
Fix matchDescriptor regexp

kenju committed:
Remove TODO

kenju committed:
DRYer regexp const

michiomochi commented on README.md:
````diff
@@ -0,0 +1,103 @@
+For more information, see http://w3c.github.io/html/semantics-embedded-content.html#element-attrdef-img-srcset
+
+```jsx
+<Image
````
READMEには書かれているReact ElementがどのようなHTMLタグを吐き出すのかも書いておくと使うイメージが沸いていいかなと思いました！

kenju replied:
たしかに、追加しますね！

kenju replied:
↓こんな感じで追加しました！
https://github.com/bitjourney/react-simple-image/pull/1#issuecomment-269411150

michiomochi commented:
見ました！nitsのみですー！
LGTM 👏

gfx commented:
インターフェイスはよさそうなので、READMEをととのえたら一旦これでいいと思います。 v0.1.0 としてリリースして組み込みましょう。

今後の課題もいくつかあるので、気が向いたらやってください。

* React.PureComponent にできないか検討してみる（PureComponentとして妥当ならそのほうが速い）
* `componentWillReceiveProps` を提供する
  * 一度elementを生成したあとに与えるpropsを変更するときにこれが必要
  * たとえばマウスオーバーでsrcSetを変えようとするときなど

kenju committed:
DRYer spec inputs array const

kenju committed:
Add className to spec

kenju committed:
Update README

michiomochi commented on spec/image.spec.jsx:
```diff
@@ -0,0 +1,126 @@
+        ],
+      };
+      const html = renderToString(createElement(Image, props));
+      console.log(html);
```
[nits] `console.log`がまぎれてます 👀

kenju replied:
🙏

kenju committed:
Remove console statement

kenju commented:
レビューありがとうございます！

> 今後の課題もいくつかあるので、気が向いたらやってください。

アドバイスもらった２点は、一旦Projectsに追記しておきました。
https://github.com/bitjourney/react-simple-image/projects/1

> たとえばマウスオーバーでsrcSetを変えようとするときなど

なるほど、たしかに 🤔
